google/ggrc-core
# GGRC-3636 Related assessment resource

# Dependencies

~On hold until frontend part is finished, so we can see if this part needs any modifications. Even though this could be merged before frontend is finished, we will still wait to avoid too many PRs for small changes that might take place.~

- [ ] ~GGRC-4328~

This PR should be safe to merge before the frontend part is finished. but FE is good to verify that everything works.

# Issue description

Add optimized endpoint for related assessments

This speeds up fetching data for related assessment from 6s +  to around 0.1s on a smaller test sample.

More benchmarks will be added.

# Steps to test the changes

- Check that related assessments still work without frontend changes. pay attention to pagination and sorting of current assessments.
- Check sorting and pagination on normal tree views (due to changes in query api code)
- Check different api calls (listed below) and verify results

- possibly if the frontend part is finished verify that it related assessments still work exactly the same before any of these PR are merged and after both (frontend and backend) PRs are merged.

# Solution description

Add extra api endpoint for retrieving related assessments:

```
/api/related_assess?object_type=Control&object_id=XXX
/api/related_assess?object_type=Control&object_id=XXX&limit=from,to
/api/related_assess?object_type=Control&object_id=XXX&limit=from,to&order_by=field_name,desc
/api/related_assess?object_type=Control&object_id=XXX&limit=from,to&order_by=field_name,asc
/api/related_assess?object_type=Control&object_id=XXX&limit=from,to&order_by=field1,desc,field2,asc
```

The limit and order by elements are made to match as closely as possible the same arguments in the query api. Backend code is refactored so that the same functions for pagination can be used here and on query API.

The object_type and object_id get parameters were chosen as a majority vote on different options.

# Sanity checklist




# PR Review checklist
